File: kotlin/lesson 6/InOrderIterative.kt
// https://leetcode.com/problems/binary-tree-inorder-traversal/

class Solution {
    fun inorderTraversal(root: TreeNode?): List<Int> {
        val list = mutableListOf<Int>()
        val stack = Stack<TreeNode>()
        var current = root
        while (current != null || !stack.empty()) {
            if (current != null) {
                stack.push(current)
                current = current.left
            } else {
                current = stack.pop()
                list.add(current.`val`)
                current = current.right
            }
        }
        return list
    }
}

File: csharp/lesson 6/InOrderIterative.cs
// https://leetcode.com/problems/binary-tree-inorder-traversal/

public class Solution
{
    public IList<int> InorderTraversal(TreeNode root)
    {
        var list = new List<int>();
        var stack = new Stack<TreeNode>();
        var current = root;
        while (current != null || stack.Count > 0)
        {
            if (current != null)
            {
                stack.Push(current);
                current = current.left;
            }
            else
            {
                current = stack.Pop();
                list.Add(current.val);
                current = current.right;
            }
        }
        return list;
    }
}
